Line/iOS/LINELoginDemo/LINELoginDemo/LineAuthManager.m

```
@@ -0,0 +1,230 @@
+            } else {
+                NSString *firebaseToken = response[@"firebase_token"];
+                [wSelf authenticateWithFirebaseToken:firebaseToken];
+            }
+        }
+    }];
+}
+
+#pragma mark - Step 3: Communication with Firebase Auth SDK
+- (void)authenticateWithFirebaseToken:(NSString *)firebaseToken {
+    // STEP 3: Login to Firebase using Firebase Custom Auth token
+    [[FIRAuth auth] signInWithCustomToken:firebaseToken completion:^(FIRUser * _Nullable user, NSError * _Nullable error) {
+        if (error) {
+            [self returnSignInResult:nil error:error];
+            return;
+        }
+        
+        // Try to update
+        [self updateUserProfile:user];
+    }];
+}
+
+#pragma mark - Step 4: (Optional) Update user profile with LINE Profile
+- (void)updateUserProfile:(FIRUser *)user {
+    BOOL isProfileNeededUpdate = ([user.displayName length] == 0) || ([[user.photoURL absoluteString] length] == 0);
+    if (!isProfileNeededUpdate) {
+        [self returnSignInResult:user error:nil];
+    }
+    
+    // STEP 4: (Optional) Update user profile with LINE Profile
+    __weak typeof(self) wSelf = self;
+    [self.lineAdapter.getLineApiClient getMyProfileWithResultBlock:^(NSDictionary *aResult, NSError *aError) {
+        if (aError) {
+            // Ignore LINE profile fetch error as this step is optional
+            [self returnSignInResult:user error:nil];
+            return;
+        }
+        
+        // Update Firebase profile with LINE profile
+        FIRUserProfileChangeRequest *request = user.profileChangeRequest;
+        request.displayName = aResult[@"displayName"];
+        if (aResult[@"pictureUrl"]) {
+            request.photoURL = [NSURL URLWithString:aResult[@"pictureUrl"]];
+        }
+        [request commitChangesWithCompletion:^(NSError * _Nullable error) {
+            // Ignore profile update error as this step is optional
```

While this step is optional, ignoring errors is bad practice that we shouldn't encourage in our samples. At the very least, add a conditional block with a comment.

```objc
[request commitChangesWithCompletion:^(NSError *error) {
  if (error != nil) {
    // handle error
  }
  [wSelf returnSignInResult:user error:nil];
}];
```